From 448609ed5932cf997e7d4ccdf39fdde491675682 Mon Sep 17 00:00:00 2001
From: Thang Tran <thuutran@amperecomputing.com>
Date: Wed, 24 Jul 2024 10:18:41 +0700
Subject: [PATCH 05/10] dbus-sdr: sel: support getting sensor number from
 static cfg

Issue: When users enable dynamic-storages-only w/o dynamic-sensors, then
call the "ipmitool sel elist" command, the sensor names are different
with the sensor paths in the "ipmi_sel" logs file.

Root cause: When users call "ipmitool sel elist", ipmitool sends a
request to ipmid to get the sensor event and sensor number. Based on the
sensor number, ipmitool sends a command to get sensor information.
Currently, SEL commands are calculating sensor numbers as a dynamic
solution, but when ipmitool uses this sensor number to get sensor
information, ipmid gets sensor information from the static configuration
solution. Due to this mismatch between 2 solutions, the sensor name of
the "ipmitool sel elist" command is different with data in the
"ipmi_sel" log file.

Solution: If users enable dynamic-storages-only w/o dynamic-sensors, the
SEL commands will get the sensor number from the static configuration.

Tested:
    1. Request to call the "ipmitool sel elist" command.
    2. The list of sensor names is same as "/var/log/ipmi_sel".
       logs file.

Upstream-Status: Pending

Change-Id: I844902bd387271b259cc02d4cf1db9bf1458df63
Signed-off-by: Thang Tran <thuutran@amperecomputing.com>
---
 dbus-sdr/storagecommands.cpp | 42 +++++++++++++++++++++++++++++++++++-
 1 file changed, 41 insertions(+), 1 deletion(-)

diff --git a/dbus-sdr/storagecommands.cpp b/dbus-sdr/storagecommands.cpp
index a784d73..f2a65df 100644
--- a/dbus-sdr/storagecommands.cpp
+++ b/dbus-sdr/storagecommands.cpp
@@ -70,6 +70,29 @@ int get()
 } // namespace erase_time
 } // namespace dynamic_sensors::ipmi::sel
 
+// If users enable dynamic-storages-only w/o dynamic-sensors, the sensor's
+// information will be got from static configuration in sensor-gen.cpp.
+#ifndef FEATURE_DYNAMIC_SENSORS
+namespace ipmi
+{
+namespace sensor
+{
+extern const IdInfoMap sensors;
+
+ipmi::sensor::IdInfoMap::const_iterator
+    getSensorInfoFromStaticCfg(const std::string& sensorPath)
+{
+    return std::find_if(
+        ipmi::sensor::sensors.begin(), ipmi::sensor::sensors.end(),
+        [&sensorPath](const ipmi::sensor::IdInfoMap::value_type& findSensor) {
+        return findSensor.second.sensorPath == sensorPath;
+    });
+}
+
+} // namespace sensor
+} // namespace ipmi
+#endif // FEATURE_DYNAMIC_SENSORS
+
 namespace ipmi
 {
 
@@ -1141,10 +1164,28 @@ ipmi::RspType<uint16_t,                   // Next Record ID
                 std::cerr << "Invalid Generator ID\n";
             }
 
+#ifndef FEATURE_DYNAMIC_SENSORS
+            // Get the Sensor's information from Static configuration.
+            // The sensors are generated in sensor-gen.cpp file.
+            auto sensorInfo =
+                ipmi::sensor::getSensorInfoFromStaticCfg(sensorPath);
+            if (sensorInfo != ipmi::sensor::sensors.end())
+            {
+                sensorType = sensorInfo->second.sensorType;
+                eventType = sensorInfo->second.sensorReadingType;
+                sensorNum = sensorInfo->first;
+                // In the static configuration, maximum Sensor Number is 255,
+                // therefore, the LUN value always is 0.
+                sensorAndLun = sensorNum;
+            }
+#else
             // Get the sensor type, sensor number, and event type for the sensor
             sensorType = getSensorTypeFromPath(sensorPath);
             sensorAndLun = getSensorNumberFromPath(sensorPath);
             sensorNum = static_cast<uint8_t>(sensorAndLun);
+            eventType = getSensorEventTypeFromPath(sensorPath);
+#endif // FEATURE_DYNAMIC_SENSORS
+
             if ((generatorID & 0x0001) == 0)
             {
                 // IPMB Address
@@ -1155,7 +1196,6 @@ ipmi::RspType<uint16_t,                   // Next Record ID
                 // system software
                 generatorID |= sensorAndLun >> 8;
             }
-            eventType = getSensorEventTypeFromPath(sensorPath);
 
             // Get the event direction
             try
-- 
2.43.0

